server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+
+            // Snapshots that can be finalized after the delete operation has been removed from the cluster state
+            private final List<SnapshotsInProgress.Entry> newFinalizations = new ArrayList<>();
+
             @Override
             public ClusterState execute(ClusterState currentState) {
                 SnapshotDeletionsInProgress deletions = currentState.custom(SnapshotDeletionsInProgress.TYPE);
                 if (deletions != null) {
                     boolean changed = false;
                     if (deletions.hasDeletionsInProgress()) {
-                        assert deletions.getEntries().size() == 1 : "should have exactly one deletion in progress";
-                        SnapshotDeletionsInProgress.Entry entry = deletions.getEntries().get(0);
-                        deletions = deletions.withRemovedEntry(entry);
-                        changed = true;
+                        final SnapshotDeletionsInProgress updatedDeletions = deletions.withRemovedEntry(deleteEntry.uuid());
+                        changed = updatedDeletions != deletions;
+                        deletions = updatedDeletions;
+                        if (failure == null) {
+                            // The delete worked out so we remove the snapshot ids that it removed from the repository from queued up
+                            // delete jobs
+                            deletions = deletions.withRemovedSnapshotIds(deleteEntry.repository(), deleteEntry.getSnapshots());
+                        }
                     }
                     if (changed) {
-                        return ClusterState.builder(currentState).putCustom(SnapshotDeletionsInProgress.TYPE, deletions).build();
+                        // We removed a delete from the cluster state. If that delete caused any snapshots to not have their shard
+                        // snapshots assigned and we have to assign them now based on the new RepositoryData that resulted from the delete.
+                        final SnapshotsInProgress snapshotsInProgress = currentState.custom(SnapshotsInProgress.TYPE);
+                        // There is one special case to deal with here: If the RepositoryData passed to this method is null, then that
+                        // means that the delete operation failed to load the current RepositoryData, indicating an issue with the
+                        // repository so we fail all queued up operations.
+                        final boolean failAllQueuedOperations = repositoryData == null;
+                        if (failAllQueuedOperations) {
+                            deletions.getEntries().stream().filter(entry -> entry.repository().equals(deleteEntry.repository()))
+                                    .map(SnapshotDeletionsInProgress.Entry::uuid).forEach(deletionsToFail::add);
+                            // We failed to read repository data for this delete, it is not the job of SnapshotsService to
```

Comment:
this should only be a problem if the repository data isn't cached, right?

Reply:
Yea it's a highly unlikely scenario, that's why I figured brute force failing (just removing from CS) all things queued up for a given snapshot should be ok here.